### PR TITLE
feat(Tag): add Tag component

### PR DESCRIPTION
**Pull request by saurabhdaware**
This PR contains
- Tag component
- Documentation
- Tests

What this PR does NOT have 
- Integration with SelectInput or other Inputs (That is here - #1459)

### CODE REVIEW

File: packages/blade/src/components/Box/BaseBox/baseBoxStyles.ts [hunk 1]
```diff
@@ -175,7 +175,10 @@ const getAllProps = (
     props.borderBottomWidth ||
     props.borderTopWidth ||
     props.borderLeftWidth ||
-    props.borderRightWidth;
+    props.borderRightWidth ||
+    props.borderWidth ||
```
Comment (saurabhdaware) [inline packages/blade/src/components/Box/BaseBox/baseBoxStyles.ts:179]: `borderStyle` was not being added when we pass just `borderWidth` or `borderColor`

File: packages/blade/src/components/Tag/Tag.stories.tsx [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+          onSubmit={(e) => {
+            e.preventDefault();
+            addTag();
+          }}
+        >
+          <TextInput
+            label="Tag Label"
+            value={inputValue}
+            onChange={({ value }) => setInputValue(value ?? '')}
+            {...{ onSubmit: isReactNative() ? () => addTag() : undefined }}
```
Comment (chaitanyadeorukhkar) [inline packages/blade/src/components/Tag/Tag.stories.tsx:143]: This is allowing me to add empty tags, which shouldn't even render since children prop is mandatory. Can we do an empty check here?

File: packages/blade/src/components/Tag/Tag.tsx [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+  icon: Icon,
+  onDismiss,
+  children,
+  isDisabled,
+  testID,
+  ...styledProps
+}: TagProps): React.ReactElement | null => {
+  const [isVisible, setIsVisible] = React.useState(true);
+
+  if (!isVisible) {
```
Comment (chaitanyadeorukhkar) [inline packages/blade/src/components/Tag/Tag.tsx:55]: So we handle whether or not to show the UI of tag or null internally. The Tag component itself doesn't really get unmounted, right? If the consumer wanted to show or hide this tag conditionally, there is no way for them to do that? And we have to have faith that the consumer has implemented `onDismiss` handler

Imagine this scenario where we submit `selectedItems` to an API call:
```tsx
const selectedItems = ['Item A', 'Item B', 'Item C']

return selectedItems.map((item) => 
  <Tag 
     onDismiss={()=>removeSlectedItem(item)} // This will remove item from selectedItems array
  >
  {item}
  </Tag>
)
```
Everything is correct here and when the tag is dismissed, the array is changed, but what if the consumer doesn't implement onDismiss?

```tsx
const selectedItems = ['Item A', 'Item B', 'Item C']

return selectedItems.map((item) => 
  <Tag>
  {item}
  </Tag>
)
```
Now, on UI for the end user, selected items will change as and when they dismiss each Item  because we handle that state internally but on code, `selectedItems` array will remain as is and when the API call is made, it will send the full `selectedItems` array which may not be the end-user's intention.


tldr; we shouldn't handle this internally, let consumers handle the visibility of the tag - we can just give them a `onDismiss` callback and they can hide the dismissed item on their end. This will ensure that there will never be a mismatch for the end user. Thoughts?

Comment (saurabhdaware) [reply packages/blade/src/components/Tag/Tag.tsx:55]: Yupp thanks for pointing out. I actually implemented it this way at the start but then moved to having uncontrolled dismiss to keep it consistent with how we do it in `Alert` right now. In Alert also when we press close icon, we remove Alert and then consumer has to take care of keeping states in sync by calling state update in `onDismiss`

I also saw this is how Ant design and Atlassian design system do as well 
- https://atlassian.design/components/tag/examples#removable
- https://ant.design/components/tag

On the other hand Carbon follows what you're saying
- https://react.carbondesignsystem.com/?path=/story/components-tag--playground&args=filter:!true

I couldn't really find any reasoning for this approach but kept it this way to keep consistency with Ant design and Atlassian's design system.

Would like more opinions on this. Don't really have a strong opinion on either approaches.

Comment (chaitanyadeorukhkar) [reply packages/blade/src/components/Tag/Tag.tsx:55]: Its okay on Alert since its just a notification to the user and not necessarily something that will impact the user if the onDismiss is not handled by the developer but on Tags it can have a significant impact

File: packages/blade/src/components/Tag/Tag.tsx [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+   */
+  children: StringChildrenType;
+
+  /**
+   * Disable tag
+   */
+  isDisabled?: boolean;
+
+  /**
+   *
```
Comment (chaitanyadeorukhkar) [inline packages/blade/src/components/Tag/Tag.tsx:48]: Should we mention that this is an internal only tag and shouldn't be used in the jsdoc?

File: packages/blade/src/components/Tag/Tag.tsx [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+      flexWrap="nowrap"
+      backgroundColor="brand.gray.a100.lowContrast"
+      borderRadius="max"
+      padding={size === 'medium' ? mediumPadding : largePadding}
+      {...getStyledProps(styledProps)}
+      {...metaAttribute({ name: MetaConstants.Tag, testID })}
+      isTagFocussed={_isTagFocussed}
+    >
+      <Box display="flex" flexDirection="row" flexWrap="nowrap" alignItems="center">
+        <Desktop>
```
Comment (chaitanyadeorukhkar) [inline packages/blade/src/components/Tag/Tag.tsx:145]: Maybe use a better component name that's easier to understand? `<ShowOnDesktop>` & `<ShowOnMobile>` maybe?

File: packages/blade/src/components/Tag/Tag.tsx [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+  return (
+    <FocussableTag
+      display={(isReactNative() ? 'flex' : 'inline-flex') as never}
+      alignSelf={isReactNative() ? 'center' : undefined}
+      flexDirection="row"
+      flexWrap="nowrap"
+      backgroundColor="brand.gray.a100.lowContrast"
+      borderRadius="max"
+      padding={size === 'medium' ? mediumPadding : largePadding}
+      {...getStyledProps(styledProps)}
```
Comment (chaitanyadeorukhkar) [inline packages/blade/src/components/Tag/Tag.tsx:140]: Does adding a display prop break the UI for this? If yes, then we should add a wrapper around it?

**Review by chaitanyadeorukhkar — DISMISSED** [review]

File: packages/blade/src/components/Tag/Tag.tsx (hunk 6 above)
Comment (saurabhdaware) [reply packages/blade/src/components/Tag/Tag.tsx:193]: ops yes it was. I removed one internal wrapper and added one wrapper externally

File: packages/blade/src/components/Tag/Tag.tsx (hunk 5 above)
Comment (saurabhdaware) [reply packages/blade/src/components/Tag/Tag.tsx]: done

File: packages/blade/src/components/Tag/Tag.tsx (hunk 4 above)
Comment (saurabhdaware) [reply packages/blade/src/components/Tag/Tag.tsx:55]: done

**Review by chaitanyadeorukhkar — DISMISSED** [review]

File: packages/blade/src/components/Tag/Tag.tsx [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+   *
+   * @private
+   *
+   */
+  _isTagFocussed?: boolean;
+} & StyledPropsBlade &
+  TestID;
+
+const FocussableTag = styled(BaseBox)<{ isTagFocussed: TagProps['_isTagFocussed'] }>((props) => {
+  if (props.isTagFocussed) {
```
Comment (kamleshchandnani) [inline packages/blade/src/components/Tag/Tag.tsx:62]: need focus ring component soon now 😞

File: packages/blade/src/components/Tag/Tag.tsx [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+        padding={size === 'medium' ? mediumPadding : largePadding}
+        isTagFocussed={_isTagFocussed}
+      >
+        <ShowOnDesktop>
+          {getLeadingIcon({ size: 'small' })}
+          {getTagText({ size: 'small' })}
+          {getCloseIcon({ size: 'small' })}
+        </ShowOnDesktop>
+        <ShowOnMobile>
+          {getLeadingIcon({ size: size === 'large' ? 'medium' : 'small' })}
```
Comment (kamleshchandnani) [inline packages/blade/src/components/Tag/Tag.tsx:212]: why can't we just compute size based on screen and then pass that directly instead of 2 separate components?

File: packages/blade/src/components/Tag/__tests__/Tag.native.test.tsx [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+  });
+
+  it('should call onDismiss', () => {
+    const dismissHandler = jest.fn();
+    const { queryAllByText, getAllByLabelText } = renderWithTheme(
+      <Tag onDismiss={dismissHandler}>in:User</Tag>,
+    );
+    expect(queryAllByText('in:User')[1]).toBeTruthy();
+    fireEvent.press(getAllByLabelText('Close in:User tag')[1]);
+    expect(dismissHandler).toBeCalledWith();
```
Comment (kamleshchandnani) [inline packages/blade/src/components/Tag/__tests__/Tag.native.test.tsx:27]: assert that the `in:User` is falsy as well since it shall be dismissed?

File: packages/blade/src/components/Tag/__tests__/Tag.web.test.tsx [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+  });
+
+  it('should call onDismiss', async () => {
+    const user = userEvents.setup();
+    const dismissHandler = jest.fn();
+    const { getByRole, queryAllByText } = renderWithTheme(
+      <Tag onDismiss={dismissHandler}>in:User</Tag>,
+    );
+
+    const [_desktopTagText, mobileTagText] = queryAllByText('in:User');
```
Comment (kamleshchandnani) [inline packages/blade/src/components/Tag/__tests__/Tag.web.test.tsx:30]: what is desktop tag and mobile tag? 🤔

File: packages/blade/src/components/Tag/Tag.stories.tsx [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+        <Tag {...args} onDismiss={() => setIsTagVisible(false)}>
+          {children}
+        </Tag>
+      ) : null}
+    </Box>
+  );
+};
+
+export const Default = TagTemplate.bind({});
+Default.args = {
```
[the remark this answers is not in the document]
Comment (saurabhdaware) [reply packages/blade/src/components/Tag/Tag.stories.tsx:93]: It is for creating overall focus ring on tag when used inside Input components. Which is I marked it internal.

File: packages/blade/src/components/Tag/Tag.tsx (hunk 8 above)
Comment (saurabhdaware) [reply packages/blade/src/components/Tag/Tag.tsx:212]: That will make size dependent on client-side javascript. I wanted to avoid that since it changes height and can cause some layout shift in SSR apps. This one uses media query to switch sizes.

File: packages/blade/src/components/Tag/__tests__/Tag.native.test.tsx (hunk 9 above)
Comment (saurabhdaware) [reply packages/blade/src/components/Tag/__tests__/Tag.native.test.tsx:27]: It won't be falsy actually. We expect consumers to handle the closing state no so when dismiss is called, consumer removes the tag. From our end for unit test, just calling onDismiss handles the cases.

File: packages/blade/src/components/Tag/Tag.stories.tsx [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+        }
+
+        export default App;
+        `}
+      </Sandbox>
+    </StoryPageWrapper>
+  );
+};
+
+export default {
```
[the remark this answers is not in the document]
Comment (saurabhdaware) [reply packages/blade/src/components/Tag/Tag.stories.tsx:60]: size is consistent on figma and dev and also predictable because its same on other components so unlikely that they will miss 🤔

File: packages/blade/src/components/Tag/Tag.stories.tsx [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+          )
+        }
+
+        export default App;
+        `}
+      </Sandbox>
+    </StoryPageWrapper>
+  );
+};
+
```
[the remark this answers is not in the document]
Comment (saurabhdaware) [reply packages/blade/src/components/Tag/Tag.stories.tsx:59]: The biggest real life usecase is using it inside AutoComplete 🙈 

I have one Dropdown story where I had temporarily used Badge, I can change it to Tags.

File: packages/blade/src/components/Button/IconButton/types.ts [hunk 14]
```diff
@@ -1,12 +1,14 @@
+import type { IconButtonProps } from './IconButton';
 import type { IconComponent } from '~components/Icons';
 import type { TestID } from '~utils/types';
 import type { BladeCommonEvents } from '~components/types';
 
 export type StyledIconButtonProps = {
   icon: IconComponent;
   onClick: () => void;
-  size: 'medium' | 'large';
```
[the remark this answers is not in the document]
Comment (saurabhdaware) [reply packages/blade/src/components/Button/IconButton/types.ts:8]: done. pushing in next commit.

File: packages/blade/src/components/Tag/Tag.stories.tsx [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+export const Default = TagTemplate.bind({});
+Default.args = {
+  children: 'Unpaid',
+  icon: 'FileTextIcon',
+} as TagProps & { icon: string };
+
+export const Disabled = TagTemplate.bind({});
```
[the remark this answers is not in the document]
Comment (saurabhdaware) [reply packages/blade/src/components/Tag/Tag.stories.tsx:98]: Will have to change the focus ring on our IconButton components. Mentioned it to design here - https://www.figma.com/file/jubmQL9Z8V7881ayUD95ps?node-id=29744:568562&mode=dev#509768673

File: packages/blade/src/components/Tag/__tests__/Tag.web.test.tsx (hunk 10 above)
Comment (saurabhdaware) [reply packages/blade/src/components/Tag/__tests__/Tag.web.test.tsx:30]: There will tags of 2 sizes in DOM and we switch between them with media query (explained it [here](https://github.com/razorpay/blade/pull/1361#discussion_r1275745922))

File: packages/blade/src/components/Tag/Tag.stories.tsx (hunk 15 above)
[the remark this answers is not in the document]
Comment (kamleshchandnani) [reply packages/blade/src/components/Tag/Tag.stories.tsx:103]: so do we release it as it is?

File: packages/blade/src/components/Tag/__tests__/Tag.native.test.tsx (hunk 9 above)
Comment (kamleshchandnani) [reply packages/blade/src/components/Tag/__tests__/Tag.native.test.tsx:27]: but we shall have a story that captures hiding of tag as well right?

File: packages/blade/src/components/Tag/Tag.tsx (hunk 8 above)
Comment (kamleshchandnani) [reply packages/blade/src/components/Tag/Tag.tsx:212]: they layout shift will anyways get triggered because our typography scale is diff for mobile and desktop and that gets evaluated client side only 🤔

Comment (saurabhdaware) [reply packages/blade/src/components/Tag/Tag.tsx:212]: > they layout shift will anyways get triggered because our typography scale is diff for mobile and desktop and that gets evaluated client side only 🤔

Lollll didn't know about this. Yup then just isMobile check makes more sense.

File: packages/blade/src/components/Tag/__tests__/Tag.native.test.tsx (hunk 9 above)
Comment (saurabhdaware) [reply packages/blade/src/components/Tag/__tests__/Tag.native.test.tsx:27]: I tried this but its not really testing any additional behaviour apart from React's useState. Feels like too much mocking for this usecase in particular. In something like [Dropdown we test the usage behaviour](https://github.com/razorpay/blade/blob/master/packages/blade/src/components/Dropdown/__tests__/Dropdown.web.test.tsx#L398-L423) because it deals with our internal props as well such as `value` `onChange` etc. Which is not the case with Tag.

image

File: packages/blade/src/components/Tag/Tag.tsx (hunk 8 above)
Comment (saurabhdaware) [reply packages/blade/src/components/Tag/Tag.tsx]: updated the code to use isMobile

File: packages/blade/src/components/Tag/Tag.stories.tsx (hunk 15 above)
[the remark this answers is not in the document]
Comment (saurabhdaware) [reply packages/blade/src/components/Tag/Tag.stories.tsx:103]: I guess yes. Don't see any way to solve this without touching IconButton's focus ring. We were also discussing about increasing focus area of icon button so that will also change the focus ring styles by default.

**Review by kamleshchandnani — APPROVED** [review]
lets gooo 🚀
